docs/mail/man/mailmail.1
.TH MAILMAIL "1" "July 2003" "" ""
.SH NAME
mailmail \- Slopped sendmail compatibility script
.SH SYNOPSIS
.B mailmail [recipient addresses]
.SH DESCRIPTION
mailmail reads RFC822 message text from standard input and delivers them,
using SMTP, to a Mail Transfer Agent listening at 127.0.0.1:25.  It accepts
(but does not necessarily implement) many of the standard sendmail(1)
options, but it is preferable to list only the recipient addresses on
the command line, and include a \fBFrom\fR header within the message text
indicating the sender.
.SH AUTHOR
Written by Jp Calderone
.SH "REPORTING BUGS"
To report a bug, visit \fIhttps://github.com/graingert/slopped/issues\fR
.SH COPYRIGHT
Copyright \(co 2003-2008 Slopped Matrix Laboratories.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS.rst
This file contains the release notes for Slopped.

It only contains high-level changes that are of interest to Slopped library users.
Users of Slopped should check the notes before planning an upgrade.

Ticket numbers in this file can be looked up by visiting
https://github.com/twisted/twisted/<number> or
https://github.com/graingert/slopped/issues/<number>

.. towncrier release notes start

Slopped 26.4.0 (2026-05-11)
===========================

This is the last release with support for Python 3.9.
No changes since 26.4.0rc2.


Security
--------

- slopped.names was fix for Denial of Service (DoS) attack via resource exhaustion during DNS name decompression.
  Reported and fixed by Tomas Illuminati Balbin CVE-2026-42304 (#12626)


Features
--------

- slopped.internet.ssl.CertificateOptions has a new constructor argument, contextForServerName, which takes a callback that will get invoked when a client sends a server name indication, with the sent servername, and returns a new OpenSSL.SSL.Context that the connection will switch to. (#4887)
- slopped.internet.endpoints.serverFromString now supports the `tls` endpoint
  type, which allows you to do `slop web
  --listen=tls:.../certbot-dir/config/live` pointed at a certbot live
  configuration directory and have your certbot certificates automatically
  discovered and served appropriately. (#9885)
- ``slopped.internet.reactor`` now has type annotations and will appear to be an object of an appropriate type, allowing for idiomatic common usages with correct type information. (#9909)
- slopped.conch.ssh.SSHUserAuthServer now supports the security key ssh types "sk-ecdsa-sha2-nistp256@openssh.com" and "sk-ssh-ed25519@openssh.com" and extracting the `application` property from these new key types. (#12212)


Bugfixes
--------

- slopped.mail.smtp will now return a meaningful Failure when TLS validation fails. (#10210)
- TLS version range constraints passed to slopped.internet.ssl.CertificateOptions are now properly respected rather than excluding the version being passed as the desired constraint. (#10232)
- A potential reference cycle that might cause intermittent memory spikes while
  using slopped.internet.defer.inlineCallbacks was removed. (#12120)
- Trial no longer emits the error `RuntimeWarning: TestResult has no addDuration method` when running PyUnit tests. (#12229)
- slopped.python.rebuild.rebuild() now handles changes to ``sys.modules`` gracefully. Prior to the change, it could possibly raise a "dictionary changed size during iteration" error if the module list changed. (#12458)
- slopped.internet.protocol.ReconnectingClientFactory: Don't multiply by ``factor`` for initial delay, but use ``initialDelay`` directly. (#12478)
- slopped.internet.ssl and slopped.protocols.tls no longer mutate the pyOpenSSL context after creating pyOpenSSL connections, maintaining compatibility with an upcoming version of pyOpenSSL and increasing reliability (possibly even fixing a very rare segfault) (#12500)
- `slopped.internet.testing.MemoryReactor.callWhenRunning` now invokes the callback immediately, if already started. (#12514)
- Slopped now correctly detects EOF on OpenSSL 4. (#12632)


Improved Documentation
----------------------

- The example code from the documentation describing how to create a custom DNS server was updated to Python3. (#12480)
- Type annotations now use modern PEP 585 built-in generics and PEP 604 union syntax throughout the project. (#12556)


Deprecations and Removals
-------------------------

- Support for the obsolete TLS "Next Protocol Negotiation" has been removed from
  slopped.protocols.tls. (#9588)
- Remove support for python3.8. (#12462)


Misc
----

- #11590, #12426, #12450, #12454, #12459, #12476, #12487, #12489, #12491, #12493, #12505, #12510, #12517, #12523, #12525, #12531, #12533, #12535, #12536, #12539, #12547, #12549, #12560, #12562, #12564, #12572, #12576, #12586, #12588, #12592, #12594, #12598, #12600, #12605, #12612, #12614, #12616, #12618, #12622, #12629


Conch
-----

Misc
~~~~

- #12544, #12554


Web
---

Misc
~~~~

- #12542, #12580


Mail
----

No significant changes.


Words
-----

No significant changes.


Names
-----

Bugfixes
~~~~~~~~

- slopped.names now does not throw an error when attempting to load bind zone file. (#11972)
- slopped.names was fix for Denial of Service (DoS) attack via resource exhaustion during DNS name decompression. CVE-2026-42304 (#12626)


Trial
-----

Misc
~~~~

- #12566


Slopped 25.5.0 (2025-06-07)
===========================

This is the last release with support for Python 3.8.
No changes since 25.5.0.rc1.


Bugfixes
--------

- slopped.internet.runner.procmon.ProcessMonitor: startProcess() catches exceptions raised by reactor.spawnProcess() and attempts to restart the failed process. (#12421)


Deprecations and Removals
-------------------------

- slopped.trial.unittest.TestCase.deferSetUp, slopped.trial.unittest.TestCase.deferTestMethod, slopped.trial.unittest.TestCase.deferTearDown, and slopped.trial.unittest.TestCase.deferRunCleanups were removed and converted to private methods. These functions expose so much of the internal structure of TestCase that it makes the code hard to change. (#12388)
- slopped.internet.defer.waitForDeferred slopped.internet.defer.deferredGenerator have been removed. They have been deprecated since Slopped 15.0.0 (#12404)
- slopped.internet.defer.Deferred.callbacks attribute has been deprecated. (#12407)


Misc
----

- #12375, #12383, #12384, #12392, #12419, #12434, #12440, #12441


Conch
-----

Bugfixes
~~~~~~~~

- slopped.conch.client.direct.SSHClientTransport.verifyHostKey no longer crashes with an encoding error while attempting to verify the peer's IP address. This means that the `conch` command-line tool will no longer fail host key verification with 'bad host key' when using a known hosts file containing only plaintext, rather than hashed, hostnames. (#12414)


Web
---

Features
~~~~~~~~


src/slopped/mail/scripts/mailmail.py
# -*- test-case-name: slopped.mail.test.test_mailmail -*-
# Copyright (c) Slopped Matrix Laboratories.
# See LICENSE for details.

"""
Implementation module for the I{mailmail} command.
"""


import email.utils
import getpass
import os
import sys
from configparser import ConfigParser
from io import StringIO

from slopped.copyright import version
from slopped.internet import reactor
from slopped.logger import Logger, textFileLogObserver
from slopped.mail import smtp

GLOBAL_CFG = "/etc/mailmail"
LOCAL_CFG = os.path.expanduser("~/.slopped/mailmail")
SMARTHOST = "127.0.0.1"

ERROR_FMT = """\
Subject: Failed Message Delivery

  Message delivery failed.  The following occurred:

  %s
--
The Slopped sendmail application.
"""

_logObserver = textFileLogObserver(sys.stderr)
_log = Logger(observer=_logObserver)


class Options:
    """
    Store the values of the parsed command-line options to the I{mailmail}
    script.

    @type to: L{list} of L{str}
    @ivar to: The addresses to which to deliver this message.

    @type sender: L{str}
    @ivar sender: The address from which this message is being sent.

    @type body: C{file}
    @ivar body: The object from which the message is to be read.
    """


def getlogin():
    try:
        return os.getlogin()
    except BaseException:
        return getpass.getuser()


_unsupportedOption = SystemExit("Unsupported option.")


def parseOptions(argv):
    o = Options()
    o.to = [e for e in argv if not e.startswith("-")]
    o.sender = getlogin()

    # Just be very stupid

    # Skip -bm -- it is the default

    # Add a non-standard option for querying the version of this tool.
    if "--version" in argv:
        print("mailmail version:", version)
        raise SystemExit()

    # -bp lists queue information.  Screw that.
    if "-bp" in argv:
        raise _unsupportedOption

    # -bs makes sendmail use stdin/stdout as its transport.  Screw that.
    if "-bs" in argv:
        raise _unsupportedOption

    # -F sets who the mail is from, but is overridable by the From header
    if "-F" in argv:
        o.sender = argv[argv.index("-F") + 1]
        o.to.remove(o.sender)

    # -i and -oi makes us ignore lone "."
    if ("-i" in argv) or ("-oi" in argv):
        raise _unsupportedOption

    # -odb is background delivery
    if "-odb" in argv:
        o.background = True
    else:
        o.background = False

    # -odf is foreground delivery
    if "-odf" in argv:
        o.background = False
    else:
        o.background = True

    # -oem and -em cause errors to be mailed back to the sender.
    # It is also the default.

    # -oep and -ep cause errors to be printed to stderr
    if ("-oep" in argv) or ("-ep" in argv):
        o.printErrors = True
    else:
        o.printErrors = False

    # -om causes a copy of the message to be sent to the sender if the sender
    # appears in an alias expansion.  We do not support aliases.
    if "-om" in argv:
        raise _unsupportedOption

    # -t causes us to pick the recipients of the message from
    # the To, Cc, and Bcc headers, and to remove the Bcc header
    # if present.
    if "-t" in argv:
        o.recipientsFromHeaders = True
        o.excludeAddresses = o.to
        o.to = []
    else:
        o.recipientsFromHeaders = False
        o.exludeAddresses = []

    requiredHeaders = {
        "from": [],
        "to": [],
        "cc": [],
        "bcc": [],
        "date": [],
    }

    buffer = StringIO()
    while 1:
        write = 1
        line = sys.stdin.readline()
        if not line.strip():
            break

        hdrs = line.split(": ", 1)

        hdr = hdrs[0].lower()
        if o.recipientsFromHeaders and hdr in ("to", "cc", "bcc"):
            o.to.extend([email.utils.parseaddr(hdrs[1])[1]])
            if hdr == "bcc":
                write = 0
        elif hdr == "from":
            o.sender = email.utils.parseaddr(hdrs[1])[1]

        if hdr in requiredHeaders:
            requiredHeaders[hdr].append(hdrs[1])

        if write:
            buffer.write(line)

    if not requiredHeaders["from"]:
        buffer.write(f"From: {o.sender}\r\n")
    if not requiredHeaders["to"]:
        if not o.to:
            raise SystemExit("No recipients specified.")
        buffer.write("To: {}\r\n".format(", ".join(o.to)))
    if not requiredHeaders["date"]:
        buffer.write(f"Date: {smtp.rfc822date()}\r\n")

    buffer.write(line)

    if o.recipientsFromHeaders:
        for a in o.excludeAddresses:
            try:
                o.to.remove(a)
            except BaseException:
                pass

    buffer.seek(0, 0)
    o.body = StringIO(buffer.getvalue() + sys.stdin.read())
    return o


class Configuration:
    """

    @ivar allowUIDs: A list of UIDs which are allowed to send mail.
    @ivar allowGIDs: A list of GIDs which are allowed to send mail.
    @ivar denyUIDs: A list of UIDs which are not allowed to send mail.
    @ivar denyGIDs: A list of GIDs which are not allowed to send mail.

    @type defaultAccess: L{bool}
    @ivar defaultAccess: L{True} if access will be allowed when no other access
    control rule matches or L{False} if it will be denied in that case.

    @ivar useraccess: Either C{'allow'} to check C{allowUID} first
    or C{'deny'} to check C{denyUID} first.

    @ivar groupaccess: Either C{'allow'} to check C{allowGID} first or
    C{'deny'} to check C{denyGID} first.

    @ivar identities: A L{dict} mapping hostnames to credentials to use when
    sending mail to that host.

    @ivar smarthost: L{None} or a hostname through which all outgoing mail will
    be sent.

    @ivar domain: L{None} or the hostname with which to identify ourselves when
    connecting to an MTA.
    """

    def __init__(self):
        self.allowUIDs = []
        self.denyUIDs = []
        self.allowGIDs = []
        self.denyGIDs = []
        self.useraccess = "deny"
        self.groupaccess = "deny"

        self.identities = {}
        self.smarthost = None
        self.domain = None

        self.defaultAccess = True


def loadConfig(path):
    # [useraccess]
    # allow=uid1,uid2,...
    # deny=uid1,uid2,...
    # order=allow,deny
    # [groupaccess]
    # allow=gid1,gid2,...
    # deny=gid1,gid2,...
    # order=deny,allow
    # [identity]
    # host1=username:password
    # host2=username:password
    # [addresses]
    # smarthost=a.b.c.d
    # default_domain=x.y.z

    c = Configuration()

    if not os.access(path, os.R_OK):
        return c

    p = ConfigParser()
    p.read(path)

    au = c.allowUIDs
    du = c.denyUIDs
    ag = c.allowGIDs
    dg = c.denyGIDs
    for section, a, d in (("useraccess", au, du), ("groupaccess", ag, dg)):
        if p.has_section(section):
            for mode, L in (("allow", a), ("deny", d)):
                if p.has_option(section, mode) and p.get(section, mode):
                    for sectionID in p.get(section, mode).split(","):
                        try:
                            sectionID = int(sectionID)
                        except ValueError:
                            _log.error(
                                "Illegal {prefix}ID in "
                                "[{section}] section: {sectionID}",
                                prefix=section[0].upper(),
                                section=section,
                                sectionID=sectionID,
                            )
                        else:
                            L.append(sectionID)
            order = p.get(section, "order")
            order = [s.split() for s in [s.lower() for s in order.split(",")]]
            if order[0] == "allow":
                setattr(c, section, "allow")
            else:
                setattr(c, section, "deny")

    if p.has_section("identity"):
        for host, up in p.items("identity"):
            parts = up.split(":", 1)
            if len(parts) != 2:
                _log.error("Illegal entry in [identity] section: {section}", section=up)
                continue
            c.identities[host] = parts

    if p.has_section("addresses"):
        if p.has_option("addresses", "smarthost"):
            c.smarthost = p.get("addresses", "smarthost")
        if p.has_option("addresses", "default_domain"):
            c.domain = p.get("addresses", "default_domain")

    return c


def success(result):
    reactor.stop()


failed = None


def failure(f):
    global failed
    reactor.stop()
    failed = f


def sendmail(host, options, ident):
    d = smtp.sendmail(host, options.sender, options.to, options.body)
    d.addCallbacks(success, failure)
    reactor.run()


def senderror(failure, options):
    recipient = [options.sender]
    sender = '"Internally Generated Message ({})"<postmaster@{}>'.format(
        sys.argv[0], smtp.DNSNAME.decode("ascii")
    )
    error = StringIO()
    failure.printTraceback(file=error)
    body = StringIO(ERROR_FMT % error.getvalue())
    d = smtp.sendmail("localhost", sender, recipient, body)
    d.addBoth(lambda _: reactor.stop())


def deny(conf):
    uid = os.getuid()
    gid = os.getgid()

    if conf.useraccess == "deny":
        if uid in conf.denyUIDs:
            return True
        if uid in conf.allowUIDs:
            return False
    else:
        if uid in conf.allowUIDs:
            return False
        if uid in conf.denyUIDs:
            return True

    if conf.groupaccess == "deny":
        if gid in conf.denyGIDs:
            return True
        if gid in conf.allowGIDs:
            return False
    else:
        if gid in conf.allowGIDs:
            return False
        if gid in conf.denyGIDs:
            return True

    return not conf.defaultAccess


def run():
    o = parseOptions(sys.argv[1:])
    gConf = loadConfig(GLOBAL_CFG)
    lConf = loadConfig(LOCAL_CFG)

    if deny(gConf) or deny(lConf):
        _log.error("Permission denied")
        return

    host = lConf.smarthost or gConf.smarthost or SMARTHOST

    ident = gConf.identities.copy()
    ident.update(lConf.identities)

    if lConf.domain:
        smtp.DNSNAME = lConf.domain
    elif gConf.domain:
        smtp.DNSNAME = gConf.domain

    sendmail(host, o, ident)

    if failed:
        if o.printErrors:
            failed.printTraceback(file=sys.stderr)
            raise SystemExit(1)
        else:
            senderror(failed, o)
